### PR TITLE
Correct and centralize creation of core/SOL ranges in IWL GK sims

### PR DESCRIPTION
**Pull request by manauref**
Some of the ranges in GK IWL clopen simulations were wrong (off by 1). For sure the global core range is wrong, and it's likely the extended core and sol ranges were also wrong.

Here we correct that.

Also, to avoid numerous definitions of LCFS related quantities (i.e. the index of the cell to its left, and the core and sol ranges) throughout the app (i.e. in GK app, in species, field, sources, bflux), and thus having an increased risk for making a mistake, we centralize the location of the LCFS parameters (x_LCFS and idx_LCFS_lo) to gk_geometry, and the ranges are only created in one place (in gyrokinetic for conf- and in species for phase-space). This way other modules can simply fetch them from these places rather than creating them again. This addreses issue #711 .

Here's proof that the ranges are correct now, done with rt_gk_d3d_iwl_3x2v_p1:
```
x_LCFS was not at a cell boundary. Moved to: 9.375000000e-02

conf grid: 0:0.15,-0.0707482:0.0707482,-3.14159:3.14159 | cells= 8, 4, 8

x_LCFS = 0.09375 | idx_LCFS_lo =  5

app->global      =  1: 8, 1: 4, 1: 8
app->global_core =  1: 5, 1: 4, 1: 8
app->global_sol  =  6: 8, 1: 4, 1: 8

app->global_ext      =  0: 9, 0: 5, 0: 9
app->global_ext_core =  0: 5, 0: 5, 0: 9
app->global_ext_sol  =  6: 9, 0: 5, 0: 9

app->lower_skin_par_core  =  1: 5, 1: 4, 1: 1
app->lower_skin_par_sol   =  6: 8, 1: 4, 1: 1
app->lower_ghost_par_core =  1: 5, 1: 4, 0: 0
app->lower_ghost_par_sol  =  6: 8, 1: 4, 0: 0

app->upper_skin_par_core  =  1: 5, 1: 4, 8: 8
app->upper_skin_par_sol   =  6: 8, 1: 4, 8: 8
app->upper_ghost_par_core =  1: 5, 1: 4, 9: 9
app->upper_ghost_par_sol  =  6: 8, 1: 4, 9: 9

app->local_par_ext_core =  1: 5, 1: 4, 0: 9
```

### CODE REVIEW

**Comment by manauref** [comment]
> Can you set the has_lcfs boolean to default to false? That way that flag doesn't need to be included in every geometry struct.

I think it's set to false by default. Did you see a case in which it wasn't?

**Comment by tnbernard** [comment]
I did not see or test anything related to the current PR. I previously saw a case specific to neutral boundary conditions. So just a note of caution.

**Comment by JunoRavin** [comment]
So long as we keep doing things like this:
struct gkyl_gk_geometry_inp geometry_inp = {
...
}
in gyrokinetic.c

and then making the app side struct by value (i.e., being careful to not malloc the app input struct because with @manauref's PR on changing to moment enums, https://github.com/ammarhakim/gkylzero/pull/712, we can still put the app input to the struct on the stack), then @tnbernard this value will default to false safely. 

We will need to make sure when inner-wall-limited capabilities are added to the gyrokinetic Lua wrappers, this boolean is set to false, but @JonathanGorard and I have tried to be careful on the Lua side to set *every* value of the input struct so that we don't have to worry about this.

**Review by JunoRavin — APPROVED** [review]

**Review by tnbernard — APPROVED** [review]

File: apps/gk_species.c [hunk 1]
```diff
@@ -1691,6 +1659,50 @@ gk_species_init(struct gkyl_gk *gk_app_inp, struct gkyl_gyrokinetic_app *app, st
...
+    int len_core = idx_LCFS_lo;
+    int len_sol = gks->global.upper[0]-len_core;
+    gkyl_range_shorten_from_above(&gks->global_core, &gks->global, 0, len_core);
+    gkyl_range_shorten_from_below(&gks->global_sol , &gks->global, 0, len_sol);
+    // Same for local ranges.
+    gkyl_range_shorten_from_above(&gks->local_core , &gks->local , 0, len_core);
+    gkyl_range_shorten_from_below(&gks->local_sol  , &gks->local , 0, len_sol);
+
+    int len_core_ext = idx_LCFS_lo+1;
+    int len_sol_ext = gks->global_ext.upper[0]-len_core;
```
Comment (Antoinehoff) [inline apps/gk_species.c:1674]: Are you sure about that?
I would have said that it's 
```
int len_sol_ext = gks->global_ext.upper[0]-(len_core + 1);
```
because the extended ranges

File: apps/gk_species.c [hunk 2]
```diff
@@ -1691,6 +1659,50 @@ gk_species_init(struct gkyl_gk *gk_app_inp, struct gkyl_gyrokinetic_app *app, st
...
+    for (int e=0; e<2; e++) {
+      gkyl_range_shorten_from_above(e==0? &gks->lower_skin_par_core  : &gks->upper_skin_par_core,
+                                    e==0? &gks->lower_skin[par_dir]  : &gks->upper_skin[par_dir], 0, len_core);
+      gkyl_range_shorten_from_above(e==0? &gks->lower_ghost_par_core : &gks->upper_ghost_par_core,
+                                    e==0? &gks->lower_ghost[par_dir] : &gks->upper_ghost[par_dir], 0, len_core);
+      gkyl_range_shorten_from_below(e==0? &gks->lower_skin_par_sol   : &gks->upper_skin_par_sol,
+                                    e==0? &gks->lower_skin[par_dir]  : &gks->upper_skin[par_dir], 0, len_sol);
+      gkyl_range_shorten_from_below(e==0? &gks->lower_ghost_par_sol  : &gks->upper_ghost_par_sol,
+                                    e==0? &gks->lower_ghost[par_dir] : &gks->upper_ghost[par_dir], 0, len_sol);
+    }
```
Comment (Antoinehoff) [inline apps/gk_species.c:1692]: Here I am also a bit doubtful. If you do the drawing for the ghost ranges the length is increased by one I think.
0-1|2|3||4|5|6|7|8|9|10-11
above '-' delimits ghosts, '||' shows the separatrix.
shorten from above:
    ghost 0 : 11
    idxLCFS_m = 3
    ghost_core_up[0] = ghost_lo[0]+len_core-1;
                      = global_lo[0]+len_core-1;
                      = 0 + 3 - 1 = 3
    ghost_core = 0:2
which does not work right?

**Review by Antoinehoff — CHANGES_REQUESTED** [review]
Since it is changing indices, we should be sure it is valgrind clean also.

File: apps/gk_species.c (hunk 1 above)
Comment (manauref) [reply apps/gk_species.c:1674]: yes, I'm sure

File: apps/gk_species.c (hunk 2 above)
Comment (manauref) [reply apps/gk_species.c:1692]: see the example I just posted in the PR message. I think all the numbers are correct/as expected now

**Review by Antoinehoff — APPROVED** [review]
great job
